### PR TITLE
executor: implement ANALYZE TABLE

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @zimulala @hanfei19910905 PTAL

File: executor/executor_simple.go [hunk 1]
```diff
@@ -310,6 +315,117 @@ func (e *SimpleExec) executeSetPwd(s *ast.SetPwdStmt) error {
...
+	st, err := parser.ParseOneStmt("select * from "+tableName, "", "")
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	compiler := &Compiler{}
+	stmt, err := compiler.Compile(e.ctx, st)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	result, err := stmt.Exec(e.ctx)
```
Comment (shenli) [inline executor/executor_simple.go]: You can use util/sqlexec.RestrictedSQLExecutor

File: executor/executor_simple.go [hunk 2]
```diff
@@ -310,6 +315,117 @@ func (e *SimpleExec) executeSetPwd(s *ast.SetPwdStmt) error {
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return result, nil
+}
+
+func (e *SimpleExec) collectSamples(result ast.RecordSet) (count int64, samples []*ast.Row, err error) {
```
Comment (shenli) [inline executor/executor_simple.go]: You can add a reference for your sampling algorithm.

**Comment by shenli** [comment]
LGTM

File: executor/executor_simple.go [hunk 3]
```diff
@@ -310,6 +314,102 @@ func (e *SimpleExec) executeSetPwd(s *ast.SetPwdStmt) error {
...
+		return errors.Trace(err)
+	}
+	err = e.buildStatisticsAndSaveToKV(tn, count, samples)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// collectSamples collects sample from the result set, use Reservoir Sampling algorithm.
```
Comment (zimulala) [inline executor/executor_simple.go]: use -> using ?

File: executor/executor_simple.go [hunk 4]
```diff
@@ -310,6 +314,102 @@ func (e *SimpleExec) executeSetPwd(s *ast.SetPwdStmt) error {
...
+			return count, samples, errors.Trace(err)
+		}
+		if row == nil {
+			break
+		}
+		if len(samples) < maxSampleCount {
+			samples = append(samples, row)
+		} else {
+			idx := ran.Intn(maxSampleCount)
+			samples[idx] = row
```
Comment (hanfei19910905) [inline executor/executor_simple.go]: This sampling algo is not very good.

File: executor/executor_simple.go [hunk 5]
```diff
@@ -310,6 +314,102 @@ func (e *SimpleExec) executeSetPwd(s *ast.SetPwdStmt) error {
...
+	if len(rows) == 0 {
+		return nil
+	}
+	columnSamples := make([][]types.Datum, len(rows[0].Data))
+	for i := range columnSamples {
+		columnSamples[i] = make([]types.Datum, len(rows))
+	}
+	for j, row := range rows {
+		for i, val := range row.Data {
+			columnSamples[i][j] = val
```
Comment (hanfei19910905) [inline executor/executor_simple.go]: why not use copy ?

Comment (coocood) [inline executor/executor_simple.go]: We need to put values of a column in a single slice, but original columns are stored in a row.

File: executor/executor_simple.go (hunk 4 above)
Comment (coocood) [inline executor/executor_simple.go]: updated.

File: executor/executor_simple.go [hunk 6]
```diff
@@ -310,6 +314,105 @@ func (e *SimpleExec) executeSetPwd(s *ast.SetPwdStmt) error {
...
+		if err != nil {
+			return count, samples, errors.Trace(err)
+		}
+		if row == nil {
+			break
+		}
+		if len(samples) < maxSampleCount {
+			samples = append(samples, row)
+		} else {
+			shouldAdd := ran.Float64() < float64(maxSampleCount)/float64(count)
```
Comment (hanfei19910905) [inline executor/executor_simple.go]: why not rand a int between 0 and count -1 ?

**Comment by hanfei19910905** [comment]
LGTM

**Pull request merged**